### ISSUE TITLE
Consider ansible for infrastructure

### ISSUE THREAD

**Issue by kofalt** [scitran/scitran#75]
Ansible could clean up and largely remove a lot of the bash scripting that we currently maintain.
Advantages would include modularity and a lower bar for understanding + contribution.

We're pretty sure this is the direction to head in, but a few questions remain:

* How to bootstrap ansible (clone + run? [probably](https://docs.ansible.com/ansible/intro_installation.html#installing-the-control-machine))
* Will we still have a convenience script that runs the ansible commands?
* Ansible has the capability to ask questions up-front, but not afterwards. Acceptable?
* Any tricks we should know? (Apparently `-c local` to avoid using ssh, etc)
* How best to use roles, variables, playbooks, and other ansible concepts?

**Comment by moloney** [comment]
For whatever it is worth, I think [SaltStack](http://saltstack.com/) should also be considered. I am currently using it to setup some infrastructure and I am quite happy with it.  Overall I think it is a bit more flexible and "programmer focused".  I don't have a ton of experience using Ansible, but I did spend a fair bit of time researching this before choosing SaltStack for my own project.

**Comment by kofalt** [comment]
Hmm. I've heard good things from multiple individuals about SaltStack - in particular, that companies tend to switch to Salt for performance reasons when managing a large fleet of hosts.

I have used Ansible before, but my knowledge is rusty. We also have a few team members (at minimum @andynemzek) who have recent experience with Ansible, so that may be a factor as well.

Fitness assessment highlights (probably incomplete):

* Declarative syntax
* Trivial environment - ideally: `git clone; cd; run`
* Supports local execution and SSH
* No server requirement
* No remote agent requirement

Based on your comment I did some light searching, and found [plenty](https://www.amon.cx/blog/saltstack-review/) of [reading](https://blog.serverdensity.com/what-ive-learnt-from-using-ansible-exclusively-for-2-years/) to jump into. I don't currently know how Salt would measure up on the above points; further comments welcome!

I should note that I'm neither the happiest nor the most experienced infrastructure scripter - merely [wrangling with pip](https://github.com/scitran/scitran/blob/dcd705baf5cd8823fc12c69e394606487c8e5e5b/scripts/0-setup.sh#L54-L118) has consumed an unfortunate amount of code and time. It's also unlikely that there will be much forward activity on this until later this year, since the current scripting is arcane but functional.

My hope is that using a declarative infrastructure tool will increase the accessibility and robustness of our infrastructure, and enable usage with tools like [Packer](https://packer.io).

**Comment by moloney** [comment]
Here is my take on those points (although I may not be understanding all of them):

 * Declarative syntax - Yes, salt uses YAML like Ansible. Salt favors the use of Jinja for looping/logic instead of having a more complex DSL (I consider this a good thing). You can also do pure python instead of YAML/Jinja.
 * Trivial environment - To bootstrap my nodes I do something like " wget -O install_salt.sh https://bootstrap.saltstack.com ; sh install_salt.sh ...". I think there is some vagrant integration too, but I haven't tried it.
 * Supports local execution and SSH -  I have been using a master/minion setup, but you have those options (see https://docs.saltstack.com/en/latest/topics/tutorials/quickstart.html).
 * No server requirement - I guess you mean you don't want a master? 
 * No remote agent requirement - Not sure what this means?

In general, people seem to get hung up on the default master/minion setup used by Salt.  A lot of times I think there isn't a real issue with this setup, but if there really is a good reason to avoid it you can.

**Comment by kofalt** [comment]
Good to know, thanks! 

By not requiring a server / remote agent, I meant that it should be possible to connect to a box over SSH and provision without first having set up a remote ansible/salt/etc installation. Based on your comments it sounds like that won't be an issue.

I also like the idea of a standard logic system over ansible's DSL.
I'll look further into this as we get closer to breaking ground on the port.

**Comment by kofalt** [comment]
@gsfr mentioned @anderiv might have insight on salt; thoughts?

**Comment by anderiv** [comment]
Hey all - I haven't use Salt at all, so I can't comment on any direct personal experience comparing it to Ansible. From the reading I've done, though, I wouldn't have any problem recommending either.

Things I really like about Ansible:

- it's agentless, and functions over the pre-existing ssh infrastructure
- it seems to strike a good balance between a purely declarative CM and "just let me run a fracking shell command, OK"? I've used puppet in the past, and had to fight it at times to get it to do what I wanted. There's *much* less of that fighting with Ansible. I suspect Salt would be much better in that regard also.
- The barrier to getting started with ansible is quite low - in most cases, I've been able to help people get up to speed with it within hours, not days.

If I was having to manage hundreds or thousands of servers, I'd probably pick Salt over Ansible, due to its ability to more quickly deploy changes. In my case, though, where I'm only managing a relatively small set of heterogeneous systems, each of which has some custom config, Ansible works great.

**Comment by anderiv** [comment]
One more note - since I have no experience with Salt, the above points I made about Ansible may very well be possible with Salt as well. I was in no way trying to argue against Salt, in case my comment may have come off that way.

**Comment by kofalt** [comment]
That's exactly the kind of information I was looking for, thanks!

**Comment by moloney** [comment]
Yeah I don't have enough experience with Ansible to make any strong statements about it.  From what I do know, I think it probably does have a easier learning curve compared to Salt.

However, I disagree that the performance differences only matter with hundreds or thousands of servers. When developing your Salt/Ansible configuration you need to be able to iterate quickly. There are some comparisons online where applying a no-op set of states (everything was already configured correctly) to just a couple of servers takes minutes on Ansible and seconds on Salt.

Anyway, either option is preferable to shell scripts!

**Comment by andynemzek** [comment]
I've been using ansible for probably about a year now and can say that no-ops are generally performed fairly fast.  If you have a huge playbook, I suppose it could become problematic.  If development is the concern, just break things out into smaller roles.  That said, there is a bit of overhead for it to make the initial connection with the machine and gather stats.

If deploying to many machines is a concern, you can always use ansible pull.  It lets you run a daemon on each machine that pulls down config from a repo and applies it to itself in a distributed manner.

**Comment by gsfr** [comment]
We are primarily looking at configuring a single machine, namely our own development environment. I think @moloney has a good point. In development, you may restart your application and, in turn, "reconfigure" your environment tens of times per day. Loosing a couple of minutes, even just 30 seconds, is highly annoying.

**Comment by andynemzek** [comment]
I agree that efficiency is key, though I think we may be talking about two different scenarios.  @moloney was suggesting that the process of actually iteratively developing your config management scripts needs to be fast.  I'm saying that even if ansible may be slower in this area (I actually don't know if it is or isn't) one can split large ansible playbooks into smaller ones (or roles) and only execute one at a time as you develop them.  This keeps the iterative overhead minimal.

On the other hand, @gsfr it sounds like you are concerned about the overhead of actually running the application.  Whatever we use, I'm not sure we want to be using config management to actually run the app in dev mode...it depends on your requirements I guess.  

I can talk a little to how I've used ansible in a dev workflow in the past.  I've used ansible as a provisioner for vagrant.  When you type 'vagrant up' in your project directory, vagrant spins up a VM and runs the ansible script.  When it's done, you have a fully configured box.  You generally only have to go through this process during your initial build, or whenever configuration changes on the host.  You don't have to worry about performing it multiple times a day.  For actually running the app in dev mode, you execute some other small script that's separate from your config management scripts and can run/restart quickly.

**Comment by kofalt** [comment]
One facet of development scripting I'm pretty adamant about is two-fold:

1. Any script target always validates everything it needs to work. Every check, every launch. Period.
1. The validation step should be very fast - checking for file existence, etc.

This gets you a bunch of useful properties, such as the trivial reset of any portion of your environment's state. This has proven incessantly invaluable in practice across several projects.

Scitran currently has these properties, though a bit slower now due to the way we have to abuse pip to get it to conform. I'm happy with either tool as long as the above properties are preserved.

**Comment by andynemzek** [comment]
@kofalt I see the use in that...it's one less thing to remember to run.  Though it seems a bit heavy to have to run it every time you run the app...even live.sh takes quite a while to launch right now and it sounds like that's why?  If using config management, I imagine this would look like us shelling out to run ansible/salt/whatever and waiting for it to complete.  I wonder if there's a middle ground to be struck...for instance, you could run config scripts every time you perform an 'update'?  Where 'update' is analogous to our current live.sh script's update command.

**Comment by kofalt** [comment]
It used to be basically instant. Now we exec pip like six times (long, exasperating story).
Fixing that would eliminate the majority of launch latency.

**Comment by andynemzek** [comment]
I'm skeptical that we'd want to run any config tool at app run time due to overhead.  If that's truly a requirement, we may want to perform some benchmark testing of various tools before we choose so we know what we're getting into.

**Comment by gsfr** [comment]
FWIW, the ```scipy``` and ```numpy``` dependencies and, hence, some of the ```pip``` runs will go away once the engine can update API metadata.
